src/sinks/new_relic/sink.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    event::Event,
+    sinks::util::{
+        builder::SinkBuilderExt,
+        encoding::EncodingConfigFixed,
+        Compression, StreamSink, RequestBuilder
+    }
+};
+use vector_core::{
+    stream::{
+        BatcherSettings, DriverResponse
+    },
+    event::{
+        Finalizable, EventFinalizers
+    },
+    buffers::Acker
+};
+use async_trait::async_trait;
+use futures::stream::{
+    BoxStream, StreamExt
+};
+use std::{
+    fmt::Debug,
+    convert::TryFrom,
+    num::NonZeroUsize,
+    sync::Arc
+};
+use tower::Service;
+use super::{
+    NewRelicApi, Encoding, NewRelicCredentials, NewRelicApiModel, NewRelicApiRequest,
+    EventsApiModel, MetricsApiModel, LogsApiModel
+};
+
+#[derive(Debug)]
+pub struct NewRelicSinkError {
+    message: String
+}
+
+impl NewRelicSinkError {
+    pub fn new(msg: &str) -> Self {
+        NewRelicSinkError {
+            message: String::from(msg)
+        }
+    }
+
+    pub fn boxed(msg: &str) -> Box<Self> {
+        Box::new(
+            NewRelicSinkError {
+                message: String::from(msg)
+            }
+        )
+    }
+}
+
+impl std::fmt::Display for NewRelicSinkError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+
+impl std::error::Error for NewRelicSinkError {
+    fn description(&self) -> &str {
+        &self.message
+    }
+}
+
+impl From<std::io::Error> for NewRelicSinkError {
+    fn from(error: std::io::Error) -> Self {
+        Self::new(&error.to_string())
+    }
+}
+
+impl From<NewRelicSinkError> for std::io::Error {
+    fn from(error: NewRelicSinkError) -> Self {
+        Self::new(std::io::ErrorKind::Other, error)
+    }
+}
+
+struct NewRelicRequestBuilder {
+    encoding: EncodingConfigFixed<Encoding>,
+    compression: Compression,
+    credentials: Arc<NewRelicCredentials>
+}
+
+impl RequestBuilder<Vec<Event>> for NewRelicRequestBuilder {
+    type Metadata = (Arc<NewRelicCredentials>, usize, EventFinalizers);
+    type Events = Result<NewRelicApiModel, Self::Error>;
+    type Encoder = EncodingConfigFixed<Encoding>;
+    type Payload = Vec<u8>;
+    type Request = NewRelicApiRequest;
+    type Error = NewRelicSinkError;
+
+    fn compression(&self) -> Compression {
+        self.compression
+    }
+
+    fn encoder(&self) -> &Self::Encoder {
+        &self.encoding
+    }
+
+    fn split_input(&self, mut input: Vec<Event>) -> (Self::Metadata, Self::Events) {
+        let events_len = input.len();
+        let finalizers = input.take_finalizers();
+        let api_model = || -> Result<NewRelicApiModel, Self::Error> {
```

Comment:
What's the purpose of the lambda here? You should just be able to return the value from the match directly
```rust
let api_model = match self.credentials.api {...}
```

Reply:
It is just to be able to use the `?` operator in the `try_from`. It's a common trick when your enclosing function/method doesn't have the same result signature as the `?` called one.